internal/controllers/scheduling/op.go

```diff
@@ -0,0 +1,233 @@
+		return nil
+
+	case syn == nil:
+		o.Reason = initialSynthesisOp
+		return o
+
+	case comp.ShouldForceResynthesis():
+		o.Reason = forcedResynthesisOp
+		return o
+
+	case comp.Synthesizing():
+		return nil
+
+	case syn.ObservedCompositionGeneration != comp.Generation:
+		o.Reason = compositionModifiedOp
+		return o
+
+	case comp.ShouldIgnoreSideEffects():
+		return nil
+	}
+
+	nonDeferredInputChanges, deferredInputChanges := inputChangeCount(synth, comp.Status.InputRevisions, syn.InputRevisions)
+	if nonDeferredInputChanges > 0 && syn.Synthesized != nil {
+		o.Reason = inputModifiedOp
+		return o
+	}
+	if deferredInputChanges > 0 && syn.Synthesized != nil {
+		o.Reason = deferredInputModifiedOp
+		return o
+	}
+
+	if syn.ObservedSynthesizerGeneration > 0 && syn.ObservedSynthesizerGeneration < synth.Generation {
+		o.Reason = synthesizerModifiedOp
+		return o
+	}
+
+	return nil
+}
+
+func (o *op) Less(than *op) bool {
+	if o.Reason == synthesizerModifiedOp && than.Reason == synthesizerModifiedOp {
+		cmp := bytes.Compare(o.SynthRolloutOrderHash(), than.SynthRolloutOrderHash())
+		if cmp != 0 {
+			return cmp > 0
+		}
+	}
+
+	if o.Reason == than.Reason {
+		return o.Composition.UID < than.Composition.UID
+	}
+
+	return o.Reason < than.Reason
+}
+
+// SynthRolloutOrderHash returns a hash that represents this composition's order in the rollout of a particular synthesizer generation.
+// This mechanism maintains determinism while shuffling the rollout order of every synthesizer change.
+func (o *op) SynthRolloutOrderHash() []byte {
+	if o.synthRolloutHash == nil {
+		hash := fnv.New64()
+		fmt.Fprintf(hash, "%s:%d:%s", o.Synthesizer.UID, o.Synthesizer.Generation, o.Composition.UID)
+		o.synthRolloutHash = hash.Sum(nil)
+	}
+	return o.synthRolloutHash
+}
+
+func (o *op) HasBeenPatched(ctx context.Context, cli client.Reader, grace time.Duration) (bool, time.Duration, error) {
+	logger := logr.FromContextOrDiscard(ctx)
+
+	wait := time.Since(o.Dispatched)
+	if wait > grace {
+		logger.V(1).Info("operation cache grace period expired", "synthesisUUID", o.id)
+		return true, wait, nil
+	}
+
+	comp := &apiv1.Composition{}
+	err := cli.Get(ctx, client.ObjectKeyFromObject(o.Composition), comp)
+	if err != nil {
+		return false, 0, err
+	}
+
+	return comp.Status.CurrentSynthesis != nil && comp.Status.CurrentSynthesis.UUID == o.id.String(), wait, nil
+}
+
+func (o *op) BuildPatch() any {
+	ops := []jsonPatch{}
+
+	if o.id == uuid.Nil {
+		// defer generating the uuid until we know the op is definitely going to be dispatched
+		o.id = uuid.Must(uuid.NewRandom())
+	}
+
+	if reflect.DeepEqual(o.Composition.Status, apiv1.CompositionStatus{}) {
+		ops = append(ops,
+			jsonPatch{Op: "test", Path: "/status", Value: nil},
+			jsonPatch{Op: "add", Path: "/status", Value: map[string]any{}})
+	}
+
+	ops = append(ops, jsonPatch{Op: "test", Path: "/status/inputRevisions", Value: o.Composition.Status.InputRevisions})
```

Comment:
Comment on why this is important? I guess it goes back to synchronization? The inputs controller can write in-between the patch build and when it actually gets applied